### PR TITLE
Rule S2068: detect hard-coded passwords in appsettings.json files

### LINKED ISSUE #5427: Rule S2068: detect hard-coded passwords in appsettings.json and web.config files
Follow-up of #4298 to detect hard-coded credentials in [appsettings.json](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-5.0) and [web.config](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/web-config?view=aspnetcore-6.0) files.

[S2068](https://rules.sonarsource.com/csharp/RSPEC-2068) should raised on appsettings.json or web.config files.
It should not raised on empty Password of ConnectionStrings because it is already handled by S2115.

Noncompliant
``` json
{
  "ConnectionStrings": {
    "DefaultConnection": "Server=myServerAddress;Database=myDataBase;User Id=myUsername;Password=123"
  }
}
```

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
	<connectionStrings>
		<add name="myConnection" connectionString="Server=myServerAddress;Database=myDataBase;User Id=myUsername;Password=123" /> <!-- Noncompliant -->
	</connectionStrings>
</configuration>
```

Compliant
```json
{
  "ConnectionStrings": {
    "DefaultConnection": "Server=myServerAddress;Database=myDataBase;Integrated Security=True"
  }
}
```

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
	<connectionStrings>
		<add name="myConnection" connectionString="Server=myServerAddress;Database=myDataBase;Trusted_Connection=True" />
	</connectionStrings>
</configuration>
```

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5427

Commit `Squash S2068-webconfig` doesn't need a review

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs [hunk 1]
```diff
@@ -106,89 +121,174 @@ protected bool IsEnabled(AnalyzerOptions options)
...
+                    CheckAppSettings(c, path, json);
+                }
             }
+        }
 
-            public Action<SyntaxNodeAnalysisContext> AnalysisAction() =>
-                context =>
+        private void CheckAppSettings(CompilationAnalysisContext c, string path, JsonNode json)
+        {
+            var queue = new Queue<JsonNode>();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:166]: Extract as reusable walker?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AppSettings/DoNotHardcodeCredentials/Valid/AppSettings.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    "fine": "",
+    "name": "Server=localhost; Database=Test; User=SA; Password=Secret123", /* Noncompliant {{"password" detected here, make sure this is not a hard-coded credential.}} */
+    /* Noncompliant@+1 */
+    "name": "Server=localhost; Database=Test; User=SA; Password=Secret123",
+    "empty": "Server=localhost; Database=Test; User=SA; Password=",         /* Compliant, should not raise on empty passwords */
+    "nopwd": "Server=localhost; Database=Test; Integrated Security=True"    /* Compliant */
+  },
+  "AppSettings": {
+    "connection": "Server=localhost; Database=Test; User=SA; Password=Secret123",   /* Noncompliant */
+    "SomeUrl": "scheme://user:azerty123@domain.com"                                 /* Noncompliant {{Review this hard-coded URI, which may contain a credential.}}" */
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AppSettings/DoNotHardcodeCredentials/Valid/AppSettings.json]: This rule does not check that. It's the other rule S2115 `DatabasePasswordsShouldBeSecure` that does it (it's modified due to rename refactoring)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DatabasePasswordsShouldBeSecure.cs [hunk 3]
```diff
@@ -101,16 +101,9 @@ private void CheckAppSettings(SonarAnalysisContext context, CompilationAnalysisC
         private void CheckAppSettingJson(CompilationAnalysisContext c, string fullPath)
         {
             var appSettings = File.ReadAllText(fullPath);
-            if (appSettings.Contains("\"ConnectionStrings\""))
+            if (appSettings.Contains("\"ConnectionStrings\"") && JsonNode.FromString(appSettings) is { } json)
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/DatabasePasswordsShouldBeSecure.cs:104]: Wrong file 😄 This rule is `DoNotHardcodeCredentialsBase`, the one with string optimization is `DatabasePasswordsShouldBeSecure`

File: analyzers/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs [hunk 4]
```diff
@@ -46,6 +46,7 @@ public class SonarAnalysisContext
 
         private const string SonarProjectConfigFileName = "SonarProjectConfig.xml";
         private static readonly Regex WebConfigRegex = new Regex(@"[\\\/]web\.([^\\\/]+\.)?config$", RegexOptions.IgnoreCase);
+        private static readonly Regex AppSettingsRegex = new Regex(@"[\\\/]appsettings\.([^\\\/]+\.)?json$", RegexOptions.IgnoreCase);
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs:49]: Yes, we're sure about the absolute path, S4NET generates those. It's not "additional files" in terms of Roslyn analysis additional files, it's coming from `FilesToAnalyze.txt`.

Regex is handier in this case.

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs:49]: Duplicate thread, looks like a GitHub glitch

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs:49]: Duplicte thread, looks like a GitHub glitch

File: analyzers/src/SonarAnalyzer.Common/Json/JsonNode.cs [hunk 5]
```diff
@@ -108,7 +126,7 @@ public void Add(JsonNode value)
             GetEnumerator();
 
         private InvalidOperationException InvalidKind() =>
-            new InvalidOperationException($"Operation is not valid. Json kind is {Kind}");
+            new($"Operation is not valid. Json kind is {Kind}");
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Json/JsonNode.cs:129]: That wouldn't work with the ternary operators

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:166]: The walker is not as easy at it seemed. I'll leave it like this for now and we can discuss it here and do it in another PR.
The key problem is, that I need to visit `"value"`, but I do not want to visit `"value"` inside object `"key": "value"`. And node doesn't know it's parent.
I don't have a good idea for this stop mechanism without going into recursive descent and enqueuing in the `base.VisitObject`.

How would `Descendants' and `DescendantValues` actually work together?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs:166]: I've drafted attempt in #6201, let's continue the discussion there

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**